### ISSUE TITLE
A problem about performance

### ISSUE THREAD

**Issue by zyf0330** [xtaci/kcptun#527]
I use kcptun in business. And I face this problem. 
I have kcp-server which targets to remote address. These two servers are located in shenzhen and beijing. When many connections are opened by kcp-server, kcp-client cannot response in time because many connections closed before client reads response.
If I set kcp-server targeting to localhost nginx, and let nginx open connection to remote address, everything works normally.
So I want to know, is it exactly a performance problem of kcptun to open many connections concurrently?

**Comment by xtaci** [comment]
as I can tell from the situation, the link between kcpserver and target server is bad. it's not a performance issue in kcptun in my opinion, but you can try to adjust the `nofile` settings and tcp parameters in `sysctl.conf`

kcptun is a tool to strengthen the weakest link in the chain.

**Comment by zyf0330** [comment]
Thanks for your reply.
I put a nginx between kcp-server and remote, and this problem is solved. So I don't think it is related to `nofile` or other thing in `sysctl.conf`.
These two situations:
* I found that even though connections opened by kcp-server become more and more, cpu usage doesn't increase, just like kcp-server is not busy. At the same time, that error info starts appearing.
* When I use nginx, with connections more and more, cpu usage of nginx increases too. When it is over 100%, that error info appears.
I guess, kcptun has no ability to handle many connections which have a few latency. I do another test to prove this. I set target as localhost application server directly, and benchmark with same way, this problem doesn't appear.
So do you have another thought?

**Comment by xtaci** [comment]
kcptun can handle as many as 2^32 connections according to `smux` theoretically,  i cannot be a problem. i don't know what caused the problem occured

**Comment by zyf0330** [comment]
Maybe you misunderstand.
The problem should happen where kcp-server opens connection to remote. I inspect source code and find it uses `net.DialTimeout("tcp", config.Target, 5*time.Second)` to open connection, `smux` doesn't be involved.

**Comment by xtaci** [comment]
then it won't be a problem of kcptun ,try tunning `nofile`

**Comment by ccsexyz** [comment]
According to my understanding, kcptun-server only establishes a connection to the target at the same time. So the quality of the network between the kcptun-server and the target will affect the speed of establishing the connection.

https://github.com/xtaci/kcptun/blob/825ccca6ac202b108b5a61cd9aeb34ffd0c293a7/server/main.go#L78 

Why not dial in a new goroutine? I think it will be better.

**Comment by zyf0330** [comment]
@ccsexyz I just know Golang a little, and with your reminder, I find it is exactly what you describes. kcp-server opens one connection every time when accepting a stream. After opening successfully, it goroutine to handle it.

**Comment by xxxman2008** [comment]
没看懂你的 网络结构设计。。最好先把你的 结构说清楚。从上面的来看，好像你的 KCP 不是用于 服务器之间的 通道的。

**Comment by zyf0330** [comment]
@ccsexyz I have test about this.
I set kcp-client and kcp-server at one server.And I set two application servers used as kcp-server `target` at two locations, one is localhost and another is a remote server which latency to localhost from ping cmd is 150 ms. And I use `wrk2` to test http service latency. I see, `QPS` of previous target is several hundreds, of last one is around 5-6. This number is exactly 1s/150ms considering that opening a connection costs more time. I test third remote server which latency is 230ms, and `QPS` decreases to 3-4.
@xtaci I don't know if there are other considerations here to code like this. But it affects perf.

consideration |  
-- | --




consideration |  
-- | --
